src/h_vialib/secure/token.py

```diff
@@ -0,0 +1,150 @@
+"""JWT based tokens which can be used to create verifiable, expiring tokens."""
+
+from urllib.parse import urlparse, urlunparse
+
+import jwt
+from jwt import DecodeError, ExpiredSignatureError, InvalidSignatureError
+
+from h_vialib import Configuration
+from h_vialib.exceptions import InvalidToken, MissingToken
+from h_vialib.secure.expiry import as_expires, quantized_expiry
+
+
+class SecureToken:
+    """A standardized and simplified JWT token."""
+
+    TOKEN_ALGORITHM = "HS256"
+
+    def __init__(self, secret):
+        """Initialise a token creator.
+
+        :param secret: The secret to sign and check tokens with
+        """
+        self._secret = secret
+
+    def create(self, payload=None, expires=None, max_age=None):
+        """Create a secure token.
+
+        :param payload: Dict of information to put in the token
+        :param expires: Datetime by which this token with expire
+        :param max_age: ... or max age in seconds after which this will expire
+        :return: A JWT encoded token as a string
+
+        :raise ValueError: if neither expires nor max_age is specified
+        """
+        payload["exp"] = as_expires(expires, max_age)
+
+        return jwt.encode(payload, self._secret, self.TOKEN_ALGORITHM)
+
+    def verify(self, token):
+        """Decode a token and check for validity.
+
+        :param token: Token string to check
+        :return: The token payload if valid
+
+        :raise InvalidToken: If the token is invalid or expired
+        :raise MissingToken: If no token is provided
+        """
+        if not token:
+            raise MissingToken("Missing secure token")
+
+        try:
+            return jwt.decode(token, self._secret, self.TOKEN_ALGORITHM)
+        except InvalidSignatureError as err:
+            raise InvalidToken("Invalid secure token") from err
+        except ExpiredSignatureError as err:
+            raise InvalidToken("Expired secure token") from err
+        except DecodeError as err:
+            raise InvalidToken("Malformed secure token") from err
+
+
+class SecureURLToken(SecureToken):
+    """A secure token used for signing and checking URLs."""
```

Comment:
We don't directly use this inbetween class, so the two could be merged, _but..._ there's nothing Via specific up until this level. If we wanted to put this in another general lib, this is where the split would be.